Remove `enum_dispatch` dependency

TimDiekmann opened the pull request:
## 🌟 What is the purpose of this PR?

We currently use a crate called `enum_dispatch` which simplifies implementing trait propagation on traits. This, however, have some downsides as well, the most notable one is, that it requires circular dependencies. For a module-based project, this is fine but for moving on with [Better separation of crates](https://app.asana.com/0/1201707629991362/1201481007343142/f) (_internal_), this is a blocker. Removing `enum_dispatch` slightly increases the amount of code needed but really helps in readability and IDE inspections.

## 🚫 Blocked by

- #483

## 👣  Next steps

- Many implementations which are now implemented directly, could be simplified.

## 🔍 What does this change?

- Implement the traits derived by `enum_dispatch` directly (without changing the structure of the enumerations)
- Remove the dependency on `enum_dispatch`
- Added documentation on how to add a new package (see *Demo*)

### 📜 Does this require a change to the docs?

Our `enum_dispatch` module had a big chunk of documentation, which now is removed as it's not needed anymore. The amount of documentation needed for that module was alarming, that the usage was not straight forward (In fact it was very hard to use it because it required many imports which were not used directly but hidden in the proc macro).

## 🛡 What tests cover this?

The integration test suite will cover this

## 🎥 Demo

![image](https://user-images.githubusercontent.com/21277928/164217812-482cce93-478f-4c53-a8ba-ebe365b2a7be.png)
![image](https://user-images.githubusercontent.com/21277928/164217903-8969e4fc-8ecd-41c2-a80f-9ae12cdad934.png)
![image](https://user-images.githubusercontent.com/21277928/164217955-f7f931d1-fde6-4664-97d4-30352d082e65.png)
![image](https://user-images.githubusercontent.com/21277928/164218007-d66cefb3-946e-46cd-b699-9cbb92f63269.png)

Alfred-Mountfield reviewed (changes requested):
Can we either add a TODO, or add docs, on what this changes for when someone wants to implement a new package

I don't think we have a proper doc right now for the package-system and how to make a new package, I'm not expecting us to implement that as part of this PR, but we should at least mention the boilerplate and such and link to the Task structs etc.

TimDiekmann commented:
I added a module-level doc to the `package` module on how to create a simple state package by example. Also, I added a TODO to create a more complex example for packages, which requires the runners, which would introduce the adjustment of tasks (So basically adding one enum variant to the task and the package.js and package.py). For the TODO, I [added a task](https://app.asana.com/0/0/1202153839209941/f) (_internal_).

Alfred-Mountfield commented on packages/engine/src/simulation/package/mod.rs:
```diff
@@ -1,20 +1,178 @@
-A default collection of packages are usually used for the engine (
-see [`PackageConfig`](crate::simulation::config::PackageConfig)).
- */
+//! The central component of the HASH simulation engine is the Package system.
```
```suggestion
//! The Package System defines the building blocks of HASH Simulation, specifying the
//! stages of execution, and the types of logic that can be ran.
```

Alfred-Mountfield commented on packages/engine/src/simulation/package/mod.rs:
```diff
@@ -1,20 +1,178 @@
+//! The central component of the HASH simulation engine is the Package system.
+//!
+//! The only presumption that the engine makes about a simulation project is that it creates
+//! a set of agents backed by the Datastore, which may have their state changed as
```
Link?

TimDiekmann replied:
Added links to `Agent` and `stateful`

Alfred-Mountfield commented on packages/engine/src/simulation/package/mod.rs:
```diff
@@ -1,20 +1,178 @@
- */
+//! The central component of the HASH simulation engine is the Package system.
+//!
+//! The only presumption that the engine makes about a simulation project is that it creates
+//! a set of agents backed by the Datastore, which may have their state changed as
+//! simulation time progresses.
```
```suggestion
```suggestion
//! The only invariant that the engine expects of a simulation project is that it creates
//! a set of agents backed by the Datastore, which may have their state changed as
//! simulation time progresses.
```

Alfred-Mountfield commented on packages/engine/src/simulation/package/mod.rs:
```diff
@@ -1,20 +1,178 @@
+//! a set of agents backed by the Datastore, which may have their state changed as
+//! simulation time progresses.
+//!
+//! Using only this presumption clearly does not provide enough utility to a simulation
+//! so packages can be added on top of that to create a simulation engine which provides
+//! the user with a wide array of functionality.
```
Not sure I like this wording :D 
```suggestion
//! The rest of the engine logic itself is defined within packages, self-contained
//! implementations that can affect the simulation initialisation, logic, and outputs.
```

Alfred-Mountfield reviewed (changes requested):
Got a little carried away with trying to reword, we can figure it out with the next task, oops :D
